<!DOCTYPE html>
<html>

  <head>
    <meta charset='utf-8'>
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <meta name="description" content="MendelBase.jl : MendelBase contains the base functions of OpenMendel.">

    <link rel="stylesheet" type="text/css" media="screen" href="stylesheets/stylesheet.css">

    <title>MendelBase.jl</title>
  </head>

  <body>

    <!-- HEADER -->
    <div id="header_wrap" class="outer">
        <header class="inner">
          <a id="forkme_banner" href="https://github.com/OpenMendel/MendelBase.jl">View on GitHub</a>

          <h1 id="project_title">MendelBase.jl</h1>
          <h2 id="project_tagline">MendelBase contains the base functions of OpenMendel.</h2>

            <section id="downloads">
              <a class="zip_download_link" href="https://github.com/OpenMendel/MendelBase.jl/zipball/master">Download this project as a .zip file</a>
              <a class="tar_download_link" href="https://github.com/OpenMendel/MendelBase.jl/tarball/master">Download this project as a tar.gz file</a>
            </section>
        </header>
    </div>

    <!-- MAIN CONTENT -->
    <div id="main_content_wrap" class="outer">
      <section id="main_content" class="inner">
        <h3>
<a id="overview" class="anchor" href="#overview" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Overview</h3>

<p><a href="https://github.com/OpenMendel/MendelBase.jl">MendelBase</a> is a utility package that contains the base functions of <a href="https://openmendel.github.io">OpenMendel</a>. It provides the information needed to process keywords and read data from external files. MendelBase is necessary to run any analysis package, and you should install MendelBase first, before you install the other packages.</p>

<h3>
<a id="installation" class="anchor" href="#installation" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Installation</h3>

<p>Within Julia, use the package manager to install MendelBase:</p>

<pre><code>Pkg.clone("git@github.com:OpenMendel/MendelBase.jl.git")
</code></pre>

<p>This package supports Julia v0.4.</p>

<h3>
<a id="input-files" class="anchor" href="#input-files" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Input Files</h3>

<p>To make your genetic data available to OpenMendel, the various input files must be in a format
that OpenMendel can read. To provide OpenMendel with the information it needs to perform its analysis you must prepare the following files. Examples input files can be found in the docs subfolders of the analysis packages. (An analysis option won't always need every file type.)</p>

<ul>
<li>
<a href="#control-file">Control File</a>: Specifies the names of your data input and output files and any optional parameters (<em>keywords</em>) for the analysis. (For a list of common keywords, see <a href="#keywords-table">Keywords Table</a> below).</li>
<li>
<a href="#locus-file">Locus File</a>: Names and describes the genetic loci in your data.</li>
<li>
<a href="#pedigree-file">Pedigree File</a>: Gives information about your individuals, such as name, sex, family structure, and ancestry. This may be a PLINK FAM format file.</li>
<li>
<a href="#phenotype-file">Phenotype File</a>: Lists the available phenotypes.</li>
<li>
<a href="#snp-definition-file">SNP Definition File</a>: Defines your SNPs with information such as SNP name, chromosome, position, allele names. This may be a PLINK BIM format file.</li>
<li>
<a href="#snp-data-file">SNP Data File</a>: Holds the genotypes for your data set. This must be a standard binary PLINK BED file in SNP major format. If you have a SNP data file you must have a SNP definition file.</li>
</ul>

<h3>
<a id="control-files" class="anchor" href="#control-files" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Control files<a id="control-file"></a>
</h3>

<p>The Control file is a text file consisting of keywords and their assigned values. The format of the Control file is:</p>

<pre><code>Keyword = Keyword_Value(s)
</code></pre>

<p>Below is an example Control file to run Gamete Competition (text after "#" is ignored, and treated as comments):</p>

<pre><code>#
# Input files.
#
locus_file = gamete competition LocusFrame.txt
pedigree_file = gamete competition PedigreeFrame.txt
#
# Analysis parameters.
#
disease_status = ACE
affected_designator = 1
standard_errors = true
</code></pre>

<p>In the example above, there are five keywords. The first two specify input and output files: <em>"gamete competition LocusFrame.txt"</em>, <em>"gamete competition PedigreeFrame.txt"</em>. The last three specify analysis parameters: <em>disease_status</em>,  <em>affected_designator</em>, <em>standard_errors</em>. The text after the '=' are the keyword values: <em>ACE</em>, <em>1</em>, <em>true</em>. For keywords that can take a set of values, the values are separated by commas.</p>

<h3>
<a id="keywords" class="anchor" href="#keywords" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Keywords<a id="keywords-table"></a>
</h3>

<p>This is a list of the general OpenMendel keywords common to most analysis package. The keywords specific to an analysis package are covered in detail in the documentation for each package. The names of keywords are <em>not</em> case sensitive. (The keyword values <em>may</em> be case sensitive.)</p>

<table>
<thead>
<tr>
<th>Keyword</th>
<th>Default Value</th>
<th>Allowed Values</th>
<th>Short Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>affected_designator</td>
<td>affected</td>
<td></td>
<td>Disease phenotype designator</td>
</tr>
<tr>
<td>allele_pseudo_count</td>
<td>0.1</td>
<td>Real</td>
<td>Number of prior alleles or haplotypes</td>
</tr>
<tr>
<td>allele_separator</td>
<td>/</td>
<td>1 Character</td>
<td>Separates codominant alleles</td>
</tr>
<tr>
<td>complexity_threshold</td>
<td>5.00E+07</td>
<td>Real</td>
<td>Criterion for not analyzing pedigrees</td>
</tr>
<tr>
<td>disease_status</td>
<td></td>
<td></td>
<td>Name of column designating disease status</td>
</tr>
<tr>
<td>eliminate_genotypes</td>
<td>FALSE</td>
<td>TRUE, FALSE</td>
<td>Whether or not genotypes are eliminated</td>
</tr>
<tr>
<td>female</td>
<td>"female", "f", 'f', "2", '2', 2</td>
<td>Female designators separated by commas</td>
<td>Female designators</td>
</tr>
<tr>
<td>field_separator</td>
<td>','</td>
<td>1 Character</td>
<td>Character used to separates values in data files</td>
</tr>
<tr>
<td>genetic_map_function</td>
<td>Haldane</td>
<td>Haldane or Kosambi</td>
<td>Genetic map units</td>
</tr>
<tr>
<td>locus_file</td>
<td></td>
<td></td>
<td>Name of input locus file</td>
</tr>
<tr>
<td>lump_alleles</td>
<td>FALSE</td>
<td>TRUE, FALSE</td>
<td>Whether or not alleles are combined</td>
</tr>
<tr>
<td>male</td>
<td>"male", "m", 'm', "1", '1', 1</td>
<td>Male designators separated by commas</td>
<td>Male designators</td>
</tr>
<tr>
<td>new_pedigree_file</td>
<td></td>
<td></td>
<td>OpenMendel generated output pedigree file</td>
</tr>
<tr>
<td>ordered_allele_separator</td>
<td>| (vertical bar)</td>
<td>1 Character</td>
<td>Separates ordered codominant alleles</td>
</tr>
<tr>
<td>output_file</td>
<td>Mendel_Output.txt</td>
<td></td>
<td>Name of standard results output file</td>
</tr>
<tr>
<td>pedigree_file</td>
<td></td>
<td></td>
<td>Name of input pedigree file</td>
</tr>
<tr>
<td>phenotype_file</td>
<td></td>
<td></td>
<td>Name of input phenotype file</td>
</tr>
<tr>
<td>plink_input_basename</td>
<td></td>
<td></td>
<td>Basename of PLINK input files</td>
</tr>
<tr>
<td>plink_output_basename</td>
<td></td>
<td></td>
<td>Basename of PLINK output files</td>
</tr>
<tr>
<td>populations</td>
<td></td>
<td>Names of populations separated by commas</td>
<td>Names of populations</td>
</tr>
<tr>
<td>product_mode</td>
<td>TRUE</td>
<td>TRUE, FALSE</td>
<td></td>
</tr>
<tr>
<td>seed</td>
<td>1234</td>
<td>Integer</td>
<td>Random number generator seed</td>
</tr>
<tr>
<td>snpdata_file</td>
<td></td>
<td></td>
<td>Name of input SNP data file</td>
</tr>
<tr>
<td>snpdefinition_file</td>
<td></td>
<td></td>
<td>Name of input SNP definition_file</td>
</tr>
</tbody>
</table>

<h3>
<a id="general-data-file-format" class="anchor" href="#general-data-file-format" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>General Data File Format</h3>

<p>All data files are in text file format with one exception: if you have dense SNP data, the SNP data should be in binary files in PLINK BED file format. OpenMendel will also accept <a href="http://pngu.mgh.harvard.edu/%7Epurcell/plink/data.shtml">PLINK format</a> FAM and BIM files. All non-PLINK OpenMendel data files list data in columns, with a header row with field names for each column. The columns need not be in any fixed order. Many column headers have required, case sensitive names. The column separators can be any single character, specified in the control file with the keyword <em>field_separator</em>. The default separator is <em>comma</em>. If a data element contains the field separator (e.g. comma) it must be enclosed in double quotes, or it will be read as a new field. Missing data is blank ("") or "NA". The PLINK file missing data symbols are "0" or "-9".</p>

<h3>
<a id="pedigree-file" class="anchor" href="#pedigree-file" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Pedigree File<a id="pedigree-file"></a>
</h3>

<p>A Pedigree file is always required. OpenMendel will accept <a href="http://pngu.mgh.harvard.edu/%7Epurcell/plink/data.shtml">PLINK format</a> FAM Pedigree files, in which case the file must have the extension <em>.fam</em>. The OpenMendel Pedigree file format has a header row with the following field names (in any order): Pedigree,Person,Mother,Father,Sex,Twin. The only required column is one labeled either <em>“Individual”</em> or <em>“Person”</em>, giving the samples names. Each individual must either have <em>both</em> and mother <em>and</em> a father, or both mother and father must be blank. If the Mother and Father columns exist but both fields are missing for an individual, that individual is  treated as a founder. If the Mother and Father columns are missing, all individuals are founders. If the Pedigree column is missing, the data is treated as separate pedigrees of size one, and all individuals are unrelated. Within each pedigree, every person must have a unique name. If a set of individuals in the same pedigree all have the same (non-blank) value in the Twin field they are all considered monozygotic (MZ) siblings. If the Sex field is missing, all individuals are treated as female. If a Sex column exists but the sex is missing for an individual, that individual is  treated as female. If the analysis uses a <a href="#locus-file">Locus file</a> to describe the markers, the genotypes are contained in the Pedigree file, and the marker names in the Locus and Pedigree files must agree. If the analysis uses a <a href="#snp-definition-file">SNP Definition file</a> to describe the markers, the genotypes are contained in the <a href="#snp-data-file">SNP Data file</a>. The Pedigree file may also include optional columns such as traits and <a href="#populations">population</a> ancestries.</p>

<p><strong>Example Pedigree File</strong></p>

<pre><code>Pedigree,Person,Mother,Father,Sex,Twin,ABO,HtMeters
Clinton,Bill,,,male,0,A,1.88
Clinton,Hillary,,,female,0,O,NA
Clinton,Chelsea,Hillary,Bill,female,0,A,NA
Obama,Barack,,,male,0,O,1.85
Obama,Michelle,,,female,0,B,NA
Obama,Malia,Michelle,Barack,female,0,O,NA
Obama,Sasha,Michelle,Barack,female,0,B,NA
</code></pre>

<h3>
<a id="phenotype-file" class="anchor" href="#phenotype-file" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Phenotype File<a id="phenotype-file"></a>
</h3>

<p>Format</p>

<p><strong>Example Phenotype File</strong></p>

<pre><code>Locus,Phenotype,Genotypes
ABO,A,"A/A,A/O"
ABO,B,"B/B,B/O"
ABO,AB,"A/B"
ABO,O,"O/O"
Rh,+,"D/D,D/d"
Rh,-,"d/d"
Xg,+,"+/+,+/-"
Xg,-,"-/-"
</code></pre>

<h3>
<a id="locus-file" class="anchor" href="#locus-file" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Locus File<a id="locus-file"></a>
</h3>

<p>The Locus file defines your genetic markers if you are working with a relatively small number of markers. If you have a dense SNP map you should use a <a href="#snp-definition-file">SNP Definition file</a> to define your SNPs. The Locus file must have a header row with the field names (in any order): <em>Locus, Allele, Chromosome, Frequency</em>. You may also include locus position information with optional the fields <em>Morgans</em>, <em>FemaleMorgans</em> and <em>MaleMorgans</em>, and <em>Basepairs</em>. For each locus there will be one line per locus allele.</p>

<p><strong>Example Locus File</strong></p>

<pre><code>Locus,Allele,Chromosome,Frequency
ABO,A,9,0.64,0.27
ABO,B,9,0.34,0.06
ABO,O,9,1.35,NA
Rh,D,1,2.42,0.61
Rh,d,1,0.58,0.39
SNP,1,9,0,NA
SNP,2,9,0,NA
</code></pre>

<h3>
<a id="snp-definition-file" class="anchor" href="#snp-definition-file" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>SNP Definition File<a id="snp-definition-file"></a>
</h3>

<p>The SNP Definition file defines your SNPs with information such as SNP name, chromosome, position, and allele names. OpenMendel will accept <a href="http://pngu.mgh.harvard.edu/%7Epurcell/plink/data.shtml">PLINK format</a> BIM files, in which case the file must have the extension <em>.bim</em>. The OpenMendel SNP Definition file format has a header row with the field names: <em>SNP</em> or <em>Locus</em> (if this column is absent the SNPs are named 1,2,3..n); Chromosome (1-22,x; if column is absent assumed autosomal); Locus, CentiMorgans (if column is absent uses 0); Basepairs (if column is absent uses 0); Allele1, Allele2 = allele names. If allele names are absent OpenMendel uses "1" for Allele1 and "2" for Allele2.</p>

<p><strong>Example SNP Definition File</strong></p>

<pre><code>Chromosome,Locus,CentiMorgans,Basepairs,Allele1,Allele2
X,rs311165,0,2699555,C,A
X,rs28579419,0,2699645,T,G
X,rs2306736,0,2700027,C,T
X,rs5939320,0,2700202,A,G
X,rs4892819,0,2700613,G,A
X,rs5982853,0,2702143,C,T
X,rs73433431,0,2702698,T,C
</code></pre>

<h3>
<a id="snp-data-file" class="anchor" href="#snp-data-file" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>SNP Data File<a id="snp-data-file"></a>
</h3>

<p>The SNP Data file holds the SNP genotypes for your data set. The SNP Data file must be a standard <a href="http://pngu.mgh.harvard.edu/%7Epurcell/plink/data.shtml">PLINK format</a> binary BED file in SNP major format. If you have a SNP data file you must have a SNP definition file, but it need not be in PLINK format. (There is no example SNP Data file because the SNP Data file is a binary file, and not human readable.)</p>

<h3>
<a id="plink-basename" class="anchor" href="#plink-basename" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>PLINK Basename</h3>

<p>You may use the keyword <em>plink_input_basename</em> in the Control file to specify a name for your PLINK format files. If you use this keyword you do not need to name each PLINK file in the Control file, but you must supply each of the three relevant PLINK files, using that basename:  <em>your_basename.bed</em> (SNP data file), <em>your_basename.bim</em> (SNP definition file), and <em>your_basename.fam</em> (pedigree file).</p>

<h3>
<a id="populations" class="anchor" href="#populations" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Populations<a id="populations"></a>
</h3>

<p>If your individuals are of mixed ethnicity, and thus have fractional ancestry, OpenMendel will allow you to adjust for this in your analysis by using the <em>populations</em> keyword. Your populations will be named in your control file, fractional ancestries assigned to your individuals in the <a href="#pedigree-file">Pedigree file</a>, and population specific allele frequencies assigned to markers in your <a href="#locus-file">Locus file</a>.</p>

<p>For example, your <a href="#control-file">Control file</a> might look like this:</p>

<pre><code># Input and Output files.
#
locus_file = genedropping LocusFrame.txt
pedigree_file = genedropping PedigreeFrame.txt
phenotype_file = genedropping PhenotypeFrame.txt
new_pedigree_file = genedropping NewPedigreeFrame.txt
#
# Analysis parameters for GeneDropping option.
#
repetitions = 2
populations = European, African, Chinese
</code></pre>

<p>Your <a href="#pedigree-file">Pedigree file</a> would have a column for each population, with a fractional ancestry assigned. The fractional ancestries need not sum to 1, but will be normalized to 1 in the analysis.</p>

<p>An example <a href="#pedigree-file">Pedigree file</a> might look like this:</p>

<pre><code>Pedigree,Person,Mother,Father,Sex,Twin,European,African,Chinese,ABO
Clinton,Bill,,,male,0,1,0,0,A
Clinton,Hillary,,,female,0,1,0,0,O
Clinton,Chelsea,Hillary,Bill,female,0,NA,NA,NA,A
Obama,Barack,,,male,0,0.5,0.5,0,O
Obama,Michelle,,,female,0,0.3,0.7,0,B
Obama,Malia,Michelle,Barack,female,0,NA,NA,NA,O
Obama,Sasha,Michelle,Barack,female,0,NA,NA,NA,B
</code></pre>

<p>Your <a href="#locus-file">Locus file</a> would have, instead of one column labeled <em>Frequency</em>,  a frequency column for each population, where the header for the column is the name of that population, and loci have an allele frequency for each population. The population frequency field names are the population names specified in the control file. In the example Locus file below, the Control file would include the line <em>"populations = European,African,Chinese"</em>. The population frequencies need not sum to 1, but will be normalized to 1 in the analysis. </p>

<p>An example <a href="#locus-file">Locus file</a> might look like this:</p>

<pre><code>Locus,Allele,Chromosome,European,African,Chinese
ABO,A,9,0.27,0.18,0.19
ABO,B,9,0.06,0.11,0.17
ABO,O,9,NA,0.71,0.64
</code></pre>

<h3>
<a id="output-files" class="anchor" href="#output-files" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Output Files</h3>

<p>All OpenMendel analysis options will create a results file. If no name is for this file is specified in the <a href="#control-file">Control</a>, it will be given the default name Mendel_Output.txt. A number of OpenMendel options give you the option of creating output files, such as plot files, or new pedigree, locus, or SNP input files based on the analysis results. These newly created input files can be used in subsequent analysis runs. To create these files, the file type is specified by a keyword in the control file, and a filename may be specified.</p>

<p>For example:</p>

<pre><code>#
# Input and Output files.
#
plink_input_basename = gwas 1 data
output_file = gwas 1 Output.txt
manhattan_plot_file = ManPlotOutput_1.png
#
# Analysis parameters for GWAS option.
#
regression = linear
regression_formula = Trait ~ Sex
</code></pre>

<h3>
<a id="citation" class="anchor" href="#citation" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Citation</h3>

<p>If you use this analysis package in your research, please cite the following reference in the resulting publications:</p>

<p><em>Lange K, Papp JC, Sinsheimer JS, Sripracha R, Zhou H, Sobel EM (2013) Mendel: The Swiss army knife of genetic analysis programs. Bioinformatics 29:1568-1570.</em></p>



<h3>
<a id="acknowledgments" class="anchor" href="#acknowledgments" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Acknowledgments</h3>

<p>This project is supported by the National Institutes of Health under NIGMS awards R01GM053275 and R25GM103774 and NHGRI award R01HG006139.</p>
      </section>
    </div>

    <!-- FOOTER  -->
    <div id="footer_wrap" class="outer">
      <footer class="inner">
        <p class="copyright">MendelBase.jl maintained by <a href="https://github.com/OpenMendel">OpenMendel</a></p>
        <p>Published with <a href="https://pages.github.com">GitHub Pages</a></p>
      </footer>
    </div>

    

  </body>
</html>
