<!DOCTYPE html>
<html>

  <head>
    <meta charset='utf-8'>
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <meta name="description" content="MendelBase.jl : MendelBase contains the base functions of OpenMendel.">

    <link rel="stylesheet" type="text/css" media="screen" href="stylesheets/stylesheet.css">

    <title>MendelBase.jl</title>
  </head>

  <body>

    <!-- HEADER -->
    <div id="header_wrap" class="outer">
        <header class="inner">
          <a id="forkme_banner" href="https://github.com/OpenMendel/MendelBase.jl">View on GitHub</a>

          <h1 id="project_title">MendelBase.jl</h1>
          <h2 id="project_tagline">MendelBase contains the base functions of OpenMendel.</h2>

            <section id="downloads">
              <a class="zip_download_link" href="https://github.com/OpenMendel/MendelBase.jl/zipball/master">Download this project as a .zip file</a>
              <a class="tar_download_link" href="https://github.com/OpenMendel/MendelBase.jl/tarball/master">Download this project as a tar.gz file</a>
            </section>
        </header>
    </div>

    <!-- MAIN CONTENT -->
    <div id="main_content_wrap" class="outer">
      <section id="main_content" class="inner">
        <h3>
<a id="overview" class="anchor" href="#overview" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Overview</h3>

<p><a href="https://github.com/OpenMendel/MendelBase.jl">MendelBase</a> is a utility package that contains the base functions of <a href="https://openmendel.github.io">OpenMendel</a>. It provides the information needed to process keywords and read data from external files. MendelBase is necessary to run any OpenMendel analysis package, and you should first install MendelBase, before you install the analysis packages.</p>

<h3>
<a id="installation" class="anchor" href="#installation" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Installation</h3>

<p>Within Julia, use the package manager to install MendelBase:</p>

<pre><code>Pkg.clone("git@github.com:OpenMendel/MendelBase.jl.git")
</code></pre>

<p>This package supports Julia v0.4.</p>

<h3>
<a id="input-files" class="anchor" href="#input-files" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Input Files</h3>

<p>To make your genetic data available to OpenMendel, the various input files must be in a format
that OpenMendel can read. Listed here are all of OpenMendel's input file types. However, not every analysis option will use every type. (Examples input files can be found in the docs subfolders of each of the analysis packages.)</p>

<ul>
<li>
<a href="#control-file">Control File</a>: Specifies the names of your data input and output files and any optional parameters (<em>keywords</em>) for the analysis. (For a list of common keywords, see <a href="#keywords-table">Keywords Table</a> below).</li>
<li>
<a href="#pedigree-file">Pedigree File</a>: Gives information about your individuals, such as name, parents' names, and sex. This may be a PLINK FAM format file.</li>
<li>
<a href="#locus-file">Locus File</a>: Names and describes the genetic loci in your data.</li>
<li>
<a href="#phenotype-file">Phenotype File</a>: Lists the relationships between the genotypes and the phenotypes at the loci in the locus file.</li>
<li>
<a href="#snp-definition-file">SNP Definition File</a>: Defines your SNPs with information such as SNP name, chromosome, position, allele names. This may be a PLINK BIM format file.</li>
<li>
<a href="#snp-data-file">SNP Data File</a>: Holds the genotypes for your data set. This must be a standard binary PLINK BED file in SNP-major format. If you have a SNP data file, you must also have a SNP definition file.</li>
</ul>

<p>If an analysis option uses a SNP Data file (and thus also a SNP Definition file), then it does <em>not</em> use a Locus file (and thus it also does not use a Phenotype file). Similarly, if an analysis option uses a Locus file, then it does not use a SNP Data nor SNP Definition file.</p>

<h3>
<a id="control-files" class="anchor" href="#control-files" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Control files<a id="control-file"></a>
</h3>

<p>The Control file is a text file consisting of keywords and their assigned values. The format for each line of the Control file is:</p>

<pre><code>Keyword = Keyword_Value(s)
</code></pre>

<p>The names of keywords are <em>not</em> case sensitive, but the keyword values <em>may</em> be case sensitive. Below is an example Control file (any text after "#" is treated as comments and ignored):</p>

<pre><code># Input and Output files.
#
locus_file = gamete competition LocusFrame.txt
pedigree_file = gamete competition PedigreeFrame.txt
output_file = gamete competition Output.txt
#
# Analysis parameters for the Gamete Competition option.
#
trait = ACE
affected_designator = 1
standard_errors = true
</code></pre>

<p>In the example above, there are six keywords. The first three specify input and output filenames: <em>locus_file</em>, <em>pedigree_file</em>, and <em>output_file</em>. The last three specify analysis parameters: <em>trait</em>,  <em>affected_designator</em>, <em>standard_errors</em>. The text after the "=" are the keyword values, for example, <em>ACE</em>, <em>1</em>, and <em>true</em>. For keywords that can take a set of values, the values are separated by commas.</p>

<h3>
<a id="keywords" class="anchor" href="#keywords" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Keywords<a id="keywords-table"></a>
</h3>

<p>This is a list of the general OpenMendel keywords common to most analysis package. The keywords relevant to a specific analysis option are covered in detail in the documentation for that package. Again, the names of keywords are <em>not</em> case sensitive, but the keyword values <em>may</em> be case sensitive.</p>

<table>
<thead>
<tr>
<th>Keyword</th>
<th>Default Value</th>
<th>Allowed Values</th>
<th>Short Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>affected_designator</td>
<td>affected</td>
<td></td>
<td>Value used to designate affecteds at the trait field</td>
</tr>
<tr>
<td>allele_pseudo_count</td>
<td>0.1</td>
<td>Real</td>
<td>Count of pseudo alleles or haplotypes to use as a prior in a Bayesian method for estimating frequencies</td>
</tr>
<tr>
<td>allele_separator</td>
<td>/</td>
<td>1 Character</td>
<td>Separates alleles in unordered genotypes</td>
</tr>
<tr>
<td>complexity_threshold</td>
<td>5.00E+07</td>
<td>Real</td>
<td>Threshold at which pedigrees with greater complexity are not analyzed because the computation would take too long</td>
</tr>
<tr>
<td>eliminate_genotypes</td>
<td>FALSE</td>
<td>TRUE, FALSE</td>
<td>Indicator of whether genotypes should be eliminated</td>
</tr>
<tr>
<td>female</td>
<td>"female", "f", 'f', "2", '2', 2</td>
<td>Female designators separated by commas</td>
<td>Female designators</td>
</tr>
<tr>
<td>field_separator</td>
<td>','</td>
<td>1 Character</td>
<td>Character used to separate values in data files</td>
</tr>
<tr>
<td>genetic_map_function</td>
<td>Haldane</td>
<td>Haldane or Kosambi</td>
<td>Function for converting Morgans to recombination fractions</td>
</tr>
<tr>
<td>locus_file</td>
<td></td>
<td>Name of existing Locus file</td>
<td>Name of input Locus file</td>
</tr>
<tr>
<td>lump_alleles</td>
<td>FALSE</td>
<td>TRUE, FALSE</td>
<td>Indicator of whether alleles should be combined</td>
</tr>
<tr>
<td>male</td>
<td>"male", "m", 'm', "1", '1', 1</td>
<td>Male designators separated by commas</td>
<td>Male designators</td>
</tr>
<tr>
<td>new_pedigree_file</td>
<td></td>
<td>Filename</td>
<td>Name of file that will hold OpenMendel generated pedigree data</td>
</tr>
<tr>
<td>ordered_allele_separator</td>
<td>| (vertical bar)</td>
<td>1 Character</td>
<td>Separates alleles in ordered genotypes</td>
</tr>
<tr>
<td>output_file</td>
<td>Mendel_Output.txt</td>
<td>Filename</td>
<td>Name of file that will hold OpenMendel analysis results</td>
</tr>
<tr>
<td>pedigree_file</td>
<td></td>
<td>Name of existing Pedigree file</td>
<td>Name of input Pedigree file</td>
</tr>
<tr>
<td>phenotype_file</td>
<td></td>
<td>Name of existing Phenotype file</td>
<td>Name of input Phenotype file</td>
</tr>
<tr>
<td>plink_input_basename</td>
<td></td>
<td>Basename of existing .bed, .bim, and .fam files</td>
<td>Basename of PLINK input files</td>
</tr>
<tr>
<td>plink_output_basename</td>
<td></td>
<td></td>
<td>Basename of PLINK output files</td>
</tr>
<tr>
<td>populations</td>
<td></td>
<td>Names of populations separated by commas</td>
<td>Names of populations</td>
</tr>
<tr>
<td>product_mode</td>
<td>TRUE</td>
<td>TRUE, FALSE</td>
<td></td>
</tr>
<tr>
<td>seed</td>
<td>1234</td>
<td>Integer</td>
<td>Seed value for random number generator</td>
</tr>
<tr>
<td>snpdata_file</td>
<td></td>
<td>Name of existing SNP Data file (in .bed format)</td>
<td>Name of input SNP Data file</td>
</tr>
<tr>
<td>snpdefinition_file</td>
<td></td>
<td>Name of existing SNP Definition file</td>
<td>Name of input SNP Definition file</td>
</tr>
<tr>
<td>standard_errors</td>
<td>FALSE</td>
<td>TRUE, FALSE</td>
<td>Indicator whether standard errors should be output</td>
</tr>
<tr>
<td>trait</td>
<td></td>
<td></td>
<td>Name of trait column in pedigree file (use Trait for .fam files)</td>
</tr>
</tbody>
</table>

<h3>
<a id="general-data-file-format" class="anchor" href="#general-data-file-format" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>General Data File Format</h3>

<p>All data files are text files with one exception: if used, the SNP data file should be a binary file in PLINK BED file format. OpenMendel will accept <a href="http://pngu.mgh.harvard.edu/%7Epurcell/plink/data.shtml">PLINK format</a> FAM and BIM files. All non-PLINK OpenMendel data files list their data in columns, with a header row with names for each column. The columns need not be in any fixed order. Many column have required, case sensitive names that need to be in the header row. The column separators can be any single character, specified in the control file with the keyword <em>field_separator</em>. The default separator is a <em>comma</em>. If a data element contains the field separator (e.g., a comma) it must be enclosed in double quotes, or it will be read as multiple fields (e.g., "A/A, A/O"). Missing data is represented by a blank or the two characters NA. In PLINK files the missing data symbols are 0 or -9.</p>

<h3>
<a id="pedigree-file" class="anchor" href="#pedigree-file" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Pedigree File<a id="pedigree-file"></a>
</h3>

<p>A Pedigree file is always required. OpenMendel will accept <a href="http://pngu.mgh.harvard.edu/%7Epurcell/plink/data.shtml">PLINK format</a> FAM pedigree files, in which case the file must have the extension <em>.fam</em>. The OpenMendel pedigree file format has a header row with some of the following field names (in any order): Pedigree, Individual (or Person), Mother, Father, Sex, and Twin. The only required column is the one labeled either <em>Individual</em> or <em>Person</em> that provides the samples names.</p>

<p>If the Pedigree column is missing, all rows are treated as separate pedigrees of size one, and all individuals are unrelated. Within each pedigree, every individual must have a unique name. If the Mother and Father columns are missing, all individuals are treated as founders.  If the Mother and Father columns exist, each individual must <em>either</em> have both a named mother and father, <em>or</em> both parental name fields must be blank. If both fields are blank, that individual is treated as a founder. If the Sex field is missing, all individuals are treated as female. If a Sex column exists but the sex is missing for an individual, that individual is treated as female. If the Twin field is missing there are no monozygotic (MZ) twins in the data set. If the Twin field is present and a set of individuals in the same pedigree all have the same non-blank value in their Twin field, they are all considered MZ siblings.</p>

<p>If the analysis uses a <a href="#locus-file">Locus file</a> to describe the markers, the marker genotypes are contained in the Pedigree file, and the marker names in the Locus and Pedigree files must agree. If the analysis uses a <a href="#snp-definition-file">SNP Definition file</a> to describe the markers, the SNP genotypes are contained in the <a href="#snp-data-file">SNP Data file</a>, not the Pedigree file. The Pedigree file may also include optional columns such as traits and <a href="#populations">population</a> ancestries.</p>

<p><strong>Example Pedigree File</strong></p>

<pre><code>Pedigree,Person,Mother,Father,Sex,Twin,ABO,HtMeters
Clinton,Bill,,,male,0,A,1.88
Clinton,Hillary,,,female,0,O,NA
Clinton,Chelsea,Hillary,Bill,female,0,A,NA
Obama,Barack,,,male,0,O,1.85
Obama,Michelle,,,female,0,B,NA
Obama,Malia,Michelle,Barack,female,0,O,NA
Obama,Sasha,Michelle,Barack,female,0,B,NA
</code></pre>

<h3>
<a id="locus-file" class="anchor" href="#locus-file" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Locus File<a id="locus-file"></a>
</h3>

<p>The Locus file defines your genetic markers if you are working with a relatively small number of markers. (If you have a dense, genome-wide set of SNPs, then you should use a <a href="#snp-definition-file">SNP Definition file</a> to define your SNPs.) The Locus file must have a header row with the field names (in any order): <em>Locus</em>, <em>Allele</em>, <em>Chromosome</em>, and <em>Frequency</em>. You may also include locus position information with the optional fields <em>Morgans</em>, <em>FemaleMorgans</em> and <em>MaleMorgans</em>, and <em>Basepairs</em>. The Locus file contains one line for each allele at each locus. The lines for all the alleles for a particular locus should be contiguous.</p>

<p><strong>Example Locus File</strong></p>

<pre><code>Locus,Allele,Chromosome,Frequency
ABO,A,9,0.64,0.27
ABO,B,9,0.34,0.06
ABO,O,9,1.35,NA
Rh,D,1,2.42,0.61
Rh,d,1,0.58,0.39
SNP,1,9,0,NA
SNP,2,9,0,NA
</code></pre>

<h3>
<a id="phenotype-file" class="anchor" href="#phenotype-file" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Phenotype File<a id="phenotype-file"></a>
</h3>

<p>If there are loci in the Locus file that are non-codominant, then the relationships between the genotypes and the phenotypes for these loci are listed in the Phenotype File. The Phenotype File is in the standard table format with a header row. The header row must have the required field names <em>Locus</em>, <em>Phenotype</em>, and <em>Genotypes</em> in any order. Each line names a locus and a phenotype used at that locus. Also listed on that line are all the genotypes consistent with that phenotype, in a quoted list. For example, the line <code>ABO,A,"A/A,A/O"</code> indicates the genotypes A/A and A/O are both valid for the phenotype A at the ABO locus.</p>

<p><strong>Example Phenotype File</strong></p>

<pre><code>Locus,Phenotype,Genotypes
ABO,A,"A/A,A/O"
ABO,B,"B/B,B/O"
ABO,AB,"A/B"
ABO,O,"O/O"
Rh,+,"D/D,D/d"
Rh,-,"d/d"
Xg,+,"+/+,+/-"
Xg,-,"-/-"
</code></pre>

<h3>
<a id="snp-definition-file" class="anchor" href="#snp-definition-file" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>SNP Definition File<a id="snp-definition-file"></a>
</h3>

<p>The SNP Definition file defines the SNPs with information such as SNP name, chromosome, position, and allele names. OpenMendel will accept <a href="http://pngu.mgh.harvard.edu/%7Epurcell/plink/data.shtml">PLINK format</a> BIM files, in which case the file must have the extension <em>.bim</em>. The OpenMendel SNP Definition file format has a header row (in any order) with some of the field names: <em>SNP</em> or <em>Locus</em>, <em>Chromosome</em>, <em>CentiMorgans</em>, <em>Basepairs</em>, <em>Allele1</em>, and <em>Allele2</em> (these last two list the allele names). If neither a SNP nor Locus field is present, then the n SNPs are named 1, 2, 3, ..., n. If the Chromosome field is missing altogether or contains a missing value at some SNP, it is assumed to be autosomal. If present, the Chromosome field should have a value from 1 to 22 or the value X. If the CentiMorgan or Basepair fields are missing, they take the value 0. If Allele1 is missing it is assigned the value "1"; a missing Allele2 becomes "2".</p>

<p><strong>Example SNP Definition File</strong></p>

<pre><code>Chromosome,SNP,CentiMorgans,Basepairs,Allele1,Allele2
X,rs311165,0,2699555,C,A
X,rs28579419,0,2699645,T,G
X,rs2306736,0,2700027,C,T
X,rs5939320,0,2700202,A,G
X,rs4892819,0,2700613,G,A
X,rs5982853,0,2702143,C,T
X,rs73433431,0,2702698,T,C
</code></pre>

<h3>
<a id="snp-data-file" class="anchor" href="#snp-data-file" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>SNP Data File<a id="snp-data-file"></a>
</h3>

<p>The SNP Data file holds the SNP genotypes. The SNP Data file must be a standard <a href="http://pngu.mgh.harvard.edu/%7Epurcell/plink/data.shtml">PLINK format</a> binary BED file in SNP-major format. If you have a SNP data file, you must have a SNP definition file, but it need not be in PLINK format. (We do not list here an example SNP Data file because they are binary files, and thus not human readable.)</p>

<h3>
<a id="plink-basename" class="anchor" href="#plink-basename" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>PLINK Basename</h3>

<p>You may use the keyword <em>plink_input_basename</em> in the Control file to specify a name for your PLINK format files. If you use this keyword you do not need to name each PLINK file in the Control file, but you must supply each of the three relevant PLINK files, using that basename:  <em>your_basename.bed</em> (SNP data file), <em>your_basename.bim</em> (SNP definition file), and <em>your_basename.fam</em> (pedigree file).</p>

<h3>
<a id="populations" class="anchor" href="#populations" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Populations<a id="populations"></a>
</h3>

<p>If your individuals are of mixed ethnicity, and thus have fractional ancestry, OpenMendel will allow you to adjust for this in your analysis by using the <em>populations</em> keyword. The relevant populations should be named in your control file, fractional ancestries should be assigned to the individuals in the <a href="#pedigree-file">Pedigree file</a>, and population specific allele frequencies should be assigned to the markers in the <a href="#locus-file">Locus file</a>.</p>

<p>For example, the <a href="#control-file">Control file</a> might be:</p>

<pre><code># Input and Output files.
#
locus_file = genedropping LocusFrame.txt
pedigree_file = genedropping PedigreeFrame.txt
phenotype_file = genedropping PhenotypeFrame.txt
new_pedigree_file = genedropping NewPedigreeFrame.txt
#
# Analysis parameters for the Gene Dropping option.
#
repetitions = 2
populations = European, African, Chinese
</code></pre>

<p>The <a href="#pedigree-file">Pedigree file</a> should have a column for each population, with fractional ancestries assigned at each individual. The fractional ancestries need not sum to 1, but will be normalized to 1 in the analysis. An example <a href="#pedigree-file">Pedigree file</a> is:</p>

<pre><code>Pedigree,Person,Mother,Father,Sex,Twin,European,African,Chinese,ABO
Clinton,Bill,,,male,0,1,0,0,A
Clinton,Hillary,,,female,0,1,0,0,O
Clinton,Chelsea,Hillary,Bill,female,0,NA,NA,NA,A
Obama,Barack,,,male,0,0.5,0.5,0,O
Obama,Michelle,,,female,0,0.3,0.7,0,B
Obama,Malia,Michelle,Barack,female,0,NA,NA,NA,O
Obama,Sasha,Michelle,Barack,female,0,NA,NA,NA,B
</code></pre>

<p>The <a href="#locus-file">Locus file</a> should have, instead of one column labeled <em>Frequency</em>, a frequency column for each population, where the header for the column is the name of that population as listed in the Control file. Each row then lists for that locus the allele frequencies for each population. The population frequencies need not sum to 1, but will be normalized to 1 in the analysis. An example <a href="#locus-file">Locus file</a> is:</p>

<pre><code>Locus,Allele,Chromosome,European,African,Chinese
ABO,A,9,0.27,0.18,0.19
ABO,B,9,0.06,0.11,0.17
ABO,O,9,NA,0.71,0.64
</code></pre>

<h3>
<a id="output-files" class="anchor" href="#output-files" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Output Files</h3>

<p>All OpenMendel analysis options will create a results file. If no name for this file is specified in the <a href="#control-file">Control file</a>, it will be given the default name <em>Mendel_Output.txt</em>. A number of OpenMendel analysis options can create additional output files, such as plot files, or new pedigree files, based on the analysis. These newly created data files can be used as input files in subsequent analyses. To create these additional output files, the filename is specified by a keyword in the Control file. For example:</p>

<pre><code>#
# Input and Output files.
#
plink_input_basename = gwas 1 data
output_file = gwas 1 Output.txt
manhattan_plot_file = ManPlotOutput_1.png
new_pedigree_file = NewPedigreeFrame.txt
#
# Analysis parameters for the GWAS option.
#
regression = linear
regression_formula = Trait ~ Sex
</code></pre>

<h3>
<a id="citation" class="anchor" href="#citation" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Citation</h3>

<p>If you use this analysis package in your research, please cite the following reference in the resulting publications:</p>

<p><em>Lange K, Papp JC, Sinsheimer JS, Sripracha R, Zhou H, Sobel EM (2013) Mendel: The Swiss army knife of genetic analysis programs. Bioinformatics 29:1568-1570.</em></p>



<h3>
<a id="acknowledgments" class="anchor" href="#acknowledgments" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Acknowledgments</h3>

<p>This project is supported by the National Institutes of Health under NIGMS awards R01GM053275 and R25GM103774 and NHGRI award R01HG006139.</p>
      </section>
    </div>

    <!-- FOOTER  -->
    <div id="footer_wrap" class="outer">
      <footer class="inner">
        <p class="copyright">MendelBase.jl maintained by <a href="https://github.com/OpenMendel">OpenMendel</a></p>
        <p>Published with <a href="https://pages.github.com">GitHub Pages</a></p>
      </footer>
    </div>

    

  </body>
</html>
